### ISSUE TITLE
Missing attribute on union types

### ISSUE THREAD

**Issue by Nathan-Furnal** [astral-sh/ty#2855]
### Summary

`ty` doesn't flag potential missing attribute on union types.

```py
# example.py
d: dict[str, dict[str, str | int | list[str]]] = {
    "id1": {"name": "apple", "price": 5, "stores": ["Bali", "Washington"]},
    "id2": {"name": "banana", "price": 3, "stores": ["Brussels", "London", "Berlin"]},
}
d["id1"]["stores"].append("Paris")
```
```sh
uvx ty check example.py
```
>  All checks passed!

While,

```sh
uvx mypy --strict example.py
```
> example.py:5: error: Item "str" of "str | int | list[str]" has no attribute "append"  [union-attr]
example.py:5: error: Item "int" of "str | int | list[str]" has no attribute "append"  [union-attr]
Found 2 errors in 1 file (checked 1 source file)

```sh
uvx pyrefly check example.py
```
> ERROR Object of class `int` has no attribute `append`
Object of class `str` has no attribute `append` [missing-attribute]
 --> example.py:5:1
  |
5 | d["id1"]["stores"].append("Paris")
  | ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
 INFO 1 error

### Version

ty 0.0.17

**Comment by sharkdp** [comment]
ty is [able to infer a precise type](https://github.com/astral-sh/ruff/pull/22882) for these subscript expressions, and so it knows that `d["id1"]["stores"]` is really `list[str]` here. So there's no error when you try to `.append` to it.

If that were code that refers to *some* `dict[str, dict[str, str | int | list[str]]]`, we would issue an error:
```py
def f(d: dict[str, dict[str, str | int | list[str]]]):
    d["id1"]["stores"].append("Paris")  # unresolved-attribute
```
https://play.ty.dev/2a475861-f150-418f-b4d9-fb87c01bc4f5

**Comment by Nathan-Furnal** [comment]
Oh that's interesting, thanks!
